### PR TITLE
[Merged by Bors] - feat(Mathlib.Algebra.Polynomial.RingDivision): add Polynomial.nmem_nonZeroDivisors_iff

### PR DESCRIPTION
**Pull request by riccardobrasca**
We add `Polynomial.nmem_nonZeroDivisors_iff`, known as McCoy's theorem: a polynomial `P : R[X]` is a zerodivisor if and only if there is `a : R` such that `a ≠ 0` and `a • P = 0`.

- [x] depends on: #12958

Co-authored-by: <manelsadi12@gmail.com>

---

[Open in Gitpod](https://gitpod.io/from-referrer/)

### CODE REVIEW

File: Mathlib/Algebra/Polynomial/RingDivision.lean [hunk 1]
```diff
@@ -364,6 +364,73 @@ theorem natDegree_pos_of_monic_of_not_isUnit {a : R[X]} (hu : ¬ IsUnit a) (ha :
     0 < natDegree a :=
   natDegree_pos_iff_degree_pos.mpr <| degree_pos_of_monic_of_not_isUnit hu ha
 
+open nonZeroDivisors Nat Set in
+/-- *McCoy theorem*: a polynomial `P : R[X]` is a zerodivisor if and only if there is `a : R`
+such that `a ≠ 0` and `a • P = 0`. We follow the proof given in
+https://math.stackexchange.com/questions/83121/zero-divisor-in-rx/83171#83171. -/
+theorem nmem_nonZeroDivisors_iff (P : R[X]) : P ∉ R[X]⁰ ↔ ∃ (a : R), a ≠ 0 ∧ a • P = 0 := by
```
Comment (erdOne) [inline Mathlib/Algebra/Polynomial/RingDivision.lean:371]: Can you add `P ∈ R[X]⁰ ↔ ∀ a : R, a • P = 0 → a = 0`?

Comment (erdOne) [inline Mathlib/Algebra/Polynomial/RingDivision.lean:371]:
```suggestion
theorem nmem_nonZeroDivisors_iff {P : R[X]} : P ∉ R[X]⁰ ↔ ∃ a : R, a ≠ 0 ∧ a • P = 0 := by
```

Comment (riccardobrasca) [reply Mathlib/Algebra/Polynomial/RingDivision.lean]: Done.

File: Mathlib/Algebra/Polynomial/RingDivision.lean [hunk 2]
```diff
@@ -364,6 +364,80 @@ theorem natDegree_pos_of_monic_of_not_isUnit {a : R[X]} (hu : ¬ IsUnit a) (ha :
...
+        simp [← coeff_C_mul, this]
+      by_contra! H
+      rw [lt_iff_not_le] at this
+      refine this (le_csSup ⟨P.natDegree, fun i hi ↦ ?_⟩ H)
+      by_contra! H1
+      exact hi <| by rw [coeff_eq_zero_of_natDegree_lt H1, zero_smul]
+    omega
+
+open nonZeroDivisors in
+lemma mem_nonZeroDivisors_iff {P : R[X]} : P ∈ R[X]⁰ ↔ ∀ a : R, a • P = 0 → a = 0 := by
```
Comment (erdOne) [inline Mathlib/Algebra/Polynomial/RingDivision.lean:435]: maybe `protected` instead?

Comment (riccardobrasca) [reply Mathlib/Algebra/Polynomial/RingDivision.lean:435]: Ah yes, that's a good idea.

**Comment by erdOne** [comment]
maintainer merge

File: Mathlib/Algebra/Polynomial/RingDivision.lean [hunk 3]
```diff
@@ -364,6 +364,80 @@ theorem natDegree_pos_of_monic_of_not_isUnit {a : R[X]} (hu : ¬ IsUnit a) (ha :
...
+      by_contra! H1
+      exact hi <| by rw [coeff_eq_zero_of_natDegree_lt H1, zero_smul]
+    omega
+
+open nonZeroDivisors in
+protected lemma mem_nonZeroDivisors_iff {P : R[X]} : P ∈ R[X]⁰ ↔ ∀ a : R, a • P = 0 → a = 0 := by
+  refine ⟨fun h a ha ↦ C_eq_zero.1 <| h (C a) (smul_eq_C_mul a ▸ ha), fun h ↦  ?_⟩
+  contrapose! h
+  obtain ⟨a, ha⟩ := nmem_nonZeroDivisors_iff.1 h
+  exact ⟨a, ha.2, ha.1⟩
```
Comment (jcommelin) [inline Mathlib/Algebra/Polynomial/RingDivision.lean:439]:
```suggestion
  simpa [not_imp_not] using (nmem_nonZeroDivisors_iff (P := P)).not
```

**Comment by jcommelin** [comment]
I've been hacking on this proof a bit more. I suggest first proving
```lean
theorem McCoy (f : R[X]) (h : ∀ r : R, r • f = 0 → r = 0) :
    ∀ (g : R[X]), f * g = 0 → g = 0 := by
  intro g hg
  by_contra hg₀
  have key : ∃ i, g.leadingCoeff * f.coeff i ≠ 0 := by
    contrapose! hg₀
    rw [← leadingCoeff_eq_zero]
    apply h
    simp [Polynomial.ext_iff, hg₀]
  have hnonempty : {i | f.coeff i • g ≠ 0}.Nonempty := key.imp fun i hi ↦ by
    rw [Set.mem_setOf_eq, ← leadingCoeff_ne_zero, smul_eq_C_mul, leadingCoeff_mul']
    all_goals rwa [leadingCoeff_C, mul_comm]
  have hbdd : BddAbove {i | f.coeff i • g ≠ 0} := by
    use f.natDegree
    intro i (hi : f.coeff i • g ≠ 0)
    contrapose! hi
    rw [coeff_eq_zero_of_natDegree_lt hi, zero_smul]
  let m := g.natDegree
  let l := sSup {i | f.coeff i • g ≠ 0}
  have hl : f.coeff l • g ≠ 0 := Nat.sSup_mem hnonempty hbdd
  suffices (f.coeff l • g).natDegree < g.natDegree by
    apply hl
    apply McCoy f h (f.coeff l • g)
    rw [smul_eq_C_mul, mul_left_comm, hg, mul_zero]
  suffices f.coeff l * g.leadingCoeff = 0 by
    apply (natDegree_smul_le (f.coeff l) g).lt_of_ne
    contrapose! hl
    rwa [← leadingCoeff_eq_zero, ← coeff_natDegree, coeff_smul, hl, coeff_natDegree, smul_eq_mul]
  suffices (f * g).coeff (l + m) = f.coeff l * g.leadingCoeff by rw [← this, hg, coeff_zero]
  rw [coeff_mul]
  apply Finset.sum_eq_single (l, m) _ (by simp)
  simp only [Finset.mem_antidiagonal, ne_eq, Prod.forall, Prod.mk.injEq, not_and]
  intro i j hij H
  obtain hi|rfl|hi := lt_trichotomy i l
  · have hj : m < j := by omega
    rw [coeff_eq_zero_of_natDegree_lt hj, mul_zero]
  · omega
  · suffices f.coeff i • g = 0 by rw [← coeff_C_mul, ← smul_eq_C_mul, this, coeff_zero]
    contrapose! hi
    refine le_csSup hbdd hi
termination_by g => g.natDegree
```
using recursion.

**Comment by riccardobrasca** [comment]
@jcommelin I've modified the PR using your proof, thanks for the review!

File: Mathlib/Algebra/Polynomial/RingDivision.lean [hunk 4]
```diff
@@ -364,6 +364,51 @@ theorem natDegree_pos_of_monic_of_not_isUnit {a : R[X]} (hu : ¬ IsUnit a) (ha :
     0 < natDegree a :=
   natDegree_pos_iff_degree_pos.mpr <| degree_pos_of_monic_of_not_isUnit hu ha
 
+theorem eq_zero_of_mul_eq_zero_of_smul (P : R[X]) (h : ∀ r : R, r • P = 0 → r = 0) :
+    ∀ (Q : R[X]), P * Q = 0 → Q = 0 := by
+  intro Q hQ
```
Comment (Ruben-VandeVelde) [inline Mathlib/Algebra/Polynomial/RingDivision.lean:369]:
```suggestion
theorem eq_zero_of_mul_eq_zero_of_smul (P : R[X]) (h : ∀ r : R, r • P = 0 → r = 0)
    (Q : R[X]) (hQ : P * Q = 0) : Q = 0 := by
```

**Review by Ruben-VandeVelde — APPROVED** [review]
lgtm as well

**Comment by riccardobrasca** [comment]
@Ruben-VandeVelde I've reverted your last commit. `termination_by` does not like variables before the colon.

**Comment by Vierkantor** [comment]
It feels like the proof of `eq_zero_of_mul_eq_zero_of_smul` is slightly too complicated, but I can't figure out a way to simplify it (reordering proofs, trying `calc` blocks, etc.) So looks good to me, thanks!

bors r+

**Comment by riccardobrasca** [comment]
> It feels like the proof of `eq_zero_of_mul_eq_zero_of_smul` is slightly too complicated, but I can't figure out a way to simplify it (reordering proofs, trying `calc` blocks, etc.) So looks good to me, thanks!
> 
> bors r+

I think the proof is indeed surprisingly tricky! There is a better proof using localization, but it requires a much bigger machinery (and I am not sure it works for semirings...).

**Pull request closed without merging**